concent_api/core/tests/test_payments_report.py

```diff
@@ -0,0 +1,226 @@
+        - 30 Subtasks, and each provider has 3 Subtasks, 2 with same requestor, 1 with different one,
+          so each Client is assigned to exactly 3 Subtasks.
+        - Each Subtask has related ForcedSubtaskPaymentEvent, ForcedPaymentEvent and two CoverAdditionalVerificationEvent payments.
+        - Each Subtask has related ForcePaymentCommitted and either SubtaskResultsSettled or SubtaskResultsRejected.
+        """
+
+        with mock.patch(
+            'core.payments.report.service.get_list_of_payments',
+            side_effect=self.get_list_of_payments_mocks,
+        ) as get_list_of_payments_mock:
+            with mock.patch(
+                'core.payments.report.service.get_covered_additional_verification_costs',
+                side_effect=self.get_covered_additional_verification_costs_mocks,
+            ) as get_covered_additional_verification_costs_mock:
+                report = create_report()
+
+        # Test SCI calls.
+        self.assertEqual(get_list_of_payments_mock.call_count, self.number_of_pairs * 3)
+        self.assertEqual(get_covered_additional_verification_costs_mock.call_count, self.number_of_pairs * 2)
+
+        # Test Pairs Reports.
+        self.assertEqual(len(report.pair_reports), self.number_of_pairs)
+
+        # There should be exactly half of the pairs with two subtasks and half with one,
+        # but the order data returned from queryset is not deterministic, so we can't check odds and evens.
+        pairs_with_two_subtasks_count = 0
+
+        for pair_report in report.pair_reports:
+            self.assertIn(pair_report.provider, self.providers)
+            self.assertIn(pair_report.requestor, self.requestors)
+            # 2 for each of 2 or 1 Subtasks.
+            self.assertIn(len(pair_report.pending_responses.subtask_results_settled), [2, 4])
+
+            if len(pair_report.pending_responses.subtask_results_settled) == 2:
+                pairs_with_two_subtasks_count += 1
+
+            self.assertEqual(len(pair_report.pending_responses.subtask_results_rejected), 0)
+            # All 3 messages for pair.
+            self.assertEqual(len(pair_report.pending_responses.force_payment_committed), 3)
+            # 1 for each pair.
+            self.assertEqual(len(pair_report.payments.regular_payments), 1)
+            # 1 for each pair.
+            self.assertEqual(len(pair_report.payments.settlement_payments), 1)
+            # 1 for each of 1 or 2 Subtasks.
+            self.assertIn(len(pair_report.payments.forced_subtask_payments), [1, 2])
+            # 1 for each of 1 or 2 Subtasks.
+            self.assertIn(len(pair_report.payments.additional_verification_payments_for_requestor), [1, 2])
+            # 1 for each of 1 or 2 Subtasks.
+            self.assertIn(len(pair_report.payments.additional_verification_payments_for_provider), [1, 2])
+
+        self.assertEqual(pairs_with_two_subtasks_count, self.number_of_pairs / 2)
+
+    def _prepare_providers(self):
```

Comment:
Done

Reply:
the solution could be simplified:
```python
    def _prepare_client(self, client_type):
        for _ in range(self.number_of_clients_per_type):
            client = Client(
                public_key=base64.b64encode(ECCx(None).raw_pubkey)
            )
            client.full_clean()
            client.save()
            getattr(self, client_type).append(client)

```
although it can stay as is.

Reply:
Done